pkg/specs/pvcs.go

```diff
@@ -178,84 +184,123 @@ pvcLoop:
-				result.Initializing = append(result.Initializing, pvc.Name)
-				continue pvcLoop
+		// If some PVC is missing, all the instance PVCs are unusable
+		for _, expectedPVC := range expectedPVCs {
+			if !slices.Contains(pvcNames, expectedPVC) {
+				result.Unusable = append(result.Unusable, pvcNames...)
+				continue instancesLoop
 			}
 		}
 
-		if pvc.Annotations[PVCStatusAnnotationName] != PVCStatusReady {
-			// This PVC has not a Job nor a Pod using it, but it is not marked as PVCStatusReady
-			// we need to ignore it here
-			continue
+		// If we have more PVC than expected, the extra PVCs are unusable
+		for _, pvcName := range pvcNames {
+			if !slices.Contains(expectedPVCs, pvcName) {
+				result.Unusable = append(result.Unusable, pvcName)
+				contextLogger.Warning("found more PVC than those expected",
+					"instance", instanceName,
+					"expectedPVCs", expectedPVCs,
+					"foundPVCs", pvcNames,
+				)
+			}
 		}
 
-		// This PVC has not a Job nor a Pod using it, it's dangling
-		result.Dangling = append(result.Dangling, pvc.Name)
-	}
-
-	if !cluster.ShouldCreateWalArchiveVolume() {
-		return result
-	}
+		// From this point we only consider expected PVCs.
+		// Any extra PVC is already in the Unusable list
+		pvcNames = expectedPVCs
 
-	for _, instance := range getInstancesPods(podList) {
-		expectedPVCs := getExpectedInstancePVCNames(cluster, instance.Name)
-		var indexOfFoundPVCs []int
+		// Search for a Pod corresponding to this instance.
+		// If found, all the PVCs are Healthy
+		for idx := range podList {
+			if IsPodSpecUsingPVCs(podList[idx].Spec, pvcNames...) {
+				// We found a Pod using this PVCs so this
+				// PVCs are not dangling
+				result.Healthy = append(result.Healthy, pvcNames...)
+				continue instancesLoop
+			}
+		}
 
-		for idx, pvcName := range result.Healthy {
-			if slices.Contains(expectedPVCs, pvcName) {
-				indexOfFoundPVCs = append(indexOfFoundPVCs, idx)
+		// Search for a Job corresponding to this instance.
+		// If found, all the PVCs are Initializing
+		for idx := range jobList {
+			if IsPodSpecUsingPVCs(jobList[idx].Spec.Template.Spec, pvcNames...) {
+				// We have found a Job corresponding to this PVCs, so we
+				// are initializing them or the initialization has just completed
+				result.Initializing = append(result.Initializing, pvcNames...)
+				continue instancesLoop
 			}
 		}
 
-		switch {
-		case len(expectedPVCs) > len(indexOfFoundPVCs):
-			// we lost some pvc null them all
-			for _, index := range indexOfFoundPVCs {
-				result.Dangling = append(result.Dangling, result.Healthy[index])
-				result.Healthy = removeElementByIndex(result.Healthy, index)
+		// At this point the PVCs are dangling, so we need to check if some of them has
+		// a PVC status annotation marking it as non-ready.
+		// In that case, all the instance PVCs are unusable
+		for _, pvc := range pvcs {
+			// We ignore any PVC that is not expected
+			if !slices.Contains(expectedPVCs, pvc.Name) {
```

Comment:
Is this really possible? We already have https://github.com/cloudnative-pg/cloudnative-pg/pull/684/files#diff-fd3956fd8d840a5772b8b430eca9c8eb8b36d261a1c91e240052ae46dd2e4964R226

Reply:
That check doesn't remove the extra PVC from the list, so we need to filter them in this check, otherwise they could invalidate the whole instance.

Reply:
Here we are iterating over PVCs not PVC names because we need to check the annotations.